### DISCUSSION TITLE
Why not contribute to the official FreeMarker project?

### DISCUSSION THREAD

**Discussion by xhaggi** [freemarker/freemarker3#3] [General]
I wonder why you are actively working on this project and not contributing to the official FreeMarker project?

**Comment by xhaggi** [comment]
@revusky I forgot to mention you.

**Comment by revusky** [comment]
> I wonder why you are actively working on this project and not contributing to the official FreeMarker project?

Well, you might find some of the information in this thread from the "Apache FreeMarker" mailing list informative: https://lists.apache.org/thread/v4c2xt3sp7twxkf40156l9o9jkbnybjh

The bottom line is that the reason that there is such a thing as "Apache FreeMarker" exists in the first place is that I agree to donate a sizeable body of work to ASF back in 2014. You can certainly infer (quite correctly) that my doing this work outside of Apache is equivalent to the fact that I have decided not to donate any more code to ASF. In fact, I do not believe that my earlier code donation was a good idea and I regret having done that. As for why I regret it and thus decline to donate any more work to ASF, to answer that in detail is really outside the scope of what I can explain here. I have it on my TODO list to write a more detailed explanation of the situation and put it on a wesite or blog. Right now, perhaps the most I can say is that the reason I am not trying to contribute any work to what you call the "official" FreeMarker project is that I am not a masochist!

Now, as for the *Apache* FreeMarker project being the "official" FreeMarker project, that is one interpretation of reality, I suppose. However, my version of reality is that what I am working on is a continuation of work on the main stream of the codebase by the person who is (certainly more than any other individual by far) the original author. Regardless of that, this codebase is, by far, the most advanced version of the FreeMarker codebase.

The way I would put it is that this is the FreeMarker project. "Apache FreeMarker" is the "Apache FreeMarker" project. 

"Apache FreeMarker" has roughly the same relationship with FreeMarker as the "Special Olympics" has with the Olympics. Finally, people understand the difference and do not find it particularly confusing.

**Comment by xhaggi** [comment]
Very interesting. Thanks for the detailed explanation and the link to the mailing list.

**Comment by revusky** [comment]
> I wonder why you are actively working on this project and not contributing to the official FreeMarker project?

It occurs to me that since you started the discussion, I have written up a more detailed project history [here](https://github.com/freemarker/freemarker3/wiki/FreeMarker-Project-History).

I think that page could clarify the situation more.
